### PR TITLE
refactor: add gpa derivative and notes list logic, add relationships

### PR DESCRIPTION
**Pull request by hussalen**
- GPA From student is now a fully derivative attribute type.
- Notes list is fully defined now, with setter logic.

### CODE REVIEW

File: project/Student.cs [hunk 1]
```diff
@@ -78,12 +82,57 @@ private float ValidateGPA(float gpa)
...
+            {
+                sum += grade.GradeValue;
+            }
+            return sum / grades.Count;
+        }
+
+        public void AddGrade(Grade grade)
+        {
+            // Does this make sense?
+            if (grade.Student == this)
```
Comment (madhtl) [inline project/Student.cs:126]: checks whether grade.Student reference is already set to the current student, looks good

**Review by madhtl — APPROVED** [review]
Nice

**Comment by madhtl** [comment]
You could add the methods to the diagram, I think we need to update it also

**Review by nightfury394 — APPROVED** [review]
LGTM

**Pull request merged**